sklearn/preprocessing/knn_imputation.py

```diff
@@ -0,0 +1,253 @@
+                              mask.sum(axis=0) > (X.shape[0] *
+                                                  self.col_max_missing))))
+        X_col_means = np.ma.array(X, mask=mask).mean(axis=0).data
+
+        # Check if % missing in any row > col_max_missing
+        bad_rows = mask.sum(axis=1) > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "The following rows have more than {0}% missing values and "
+                "are not included as donor neighbors: {1}"
+                    .format(self.row_max_missing*100, np.where(bad_rows)))
+
+            # Remove rows that have more than row_max_missing % missing
+            X = X[~bad_rows, :]
+
+        # Check if sufficient neighboring samples available
+        if X.shape[0] < self.n_neighbors:
+            raise ValueError("There are only %d samples, "
+                             "but n_neighbors=%d."
+                             % (X.shape[0], self.n_neighbors))
+
+        # Instantiate NN object, get column means, and store in statistics_
+        neigh = NearestNeighbors(n_neighbors=self.n_neighbors,
+                                 metric=self.metric)
+        self.statistics_ = (neigh.fit(X), X_col_means)
+
+        return self
+
+    def transform(self, X):
+        """Impute all missing values in X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape = [n_samples, n_features]
+            The input data to complete.
+        """
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        check_is_fitted(self, 'statistics_')
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        mask = _get_mask(X, self.missing_values)
+        n_rows_X, n_cols_X = X.shape
+        row_total_missing = mask.sum(axis=1)
+
+        # Get fitted objects
+        fitted_NNObj, fitted_col_means = self.statistics_
+        fitted_X = fitted_NNObj._fit_X
+
+        # Check for excessive missingness in rows
+        bad_rows = row_total_missing > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "The following rows have more than {0}% missing values and "
+                "are imputed with column means: {1}"
+                    .format(self.row_max_missing*100, np.where(bad_rows)))
+            X_bad = X[bad_rows, :]
+            X = X[~bad_rows, :]
+            mask = _get_mask(X, self.missing_values)
+            row_total_missing = mask.sum(axis=1)
+
+        # Check if the X in fit() and transform() are the same
+        if X is fitted_X or X.base is fitted_X.base:
+            neighbors = fitted_NNObj.kneighbors(n_neighbors=self.n_neighbors)
+        else:
+            neighbors = fitted_NNObj.kneighbors(X,
```

Comment:
we should only do this for rows with missing values. (I know in the original genome case that might have been all of them.)